### PR TITLE
feat(api-373): importable sheepdog for gdcapi

### PR DESCRIPTION
**Pull request by ProfOak**
- Move everything into setup.py
- Fix dependency conflicts
- added .gitignore file
- added code to modify releasable flag on nodes
- uses authutils instead of hardcoded auth folder
- uses cdiserrors instead of hardcoded error file
- used cdislogging instead of cdisutils.log
- added email support
- hooked up some methods that weren't connected correctly
- re-enabled `/open`, `/review`, `/submit`, `/release` endpoints
- enabled option to use signpost instead of indexd
- add dbgap x referencer

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,4 +23,36 @@
...
+        'gdcdatamodel',
+    ],
+    dependency_links=[
+        'git+https://git@github.com/NCI-GDC/psqlgraph.git@7b5de7d56aa3159a9526940eb273579ddbf084ca#egg=psqlgraph',
+        'git+https://git@github.com/uc-cdis/cdis-python-utils.git@0.1.5#egg=cdispyutils',
+        'git+https://git@github.com/uc-cdis/userdatamodel.git@f25f5dfb356e80c45153ffd3e91b399be672a81c#egg=userdatamodel',
+        'git+https://git@github.com/uc-cdis/cdis_oauth2client.git@0.1.1#egg=cdis_oauth2client',
+        'git+https://git@github.com/NCI-GDC/cdisutils.git@a79409a0ce5071a81c6997d4ed1549c3544fbdcd#egg=cdisutils',
+        'git+https://git@github.com/uc-cdis/datadictionary.git@0.1.1#egg=gdcdictionary',
+        'git+https://git@github.com/NCI-GDC/gdcdatamodel.git@755c6d7c380b69dc36dced55700bc9e24a084db1#egg=gdcdatamodel',
```
Comment (rudyardrichter) [inline setup.py]: There are also these dependencies from the `requirements.txt` on latest master:
```
git+https://git@github.com/NCI-GDC/signpost.git@c5d499936943e71eefe2ec4b3d4ced6ac48f35c0#egg=signpost
git+https://git@github.com/uc-cdis/indexclient.git@d49134f4626b69a8ef02c189ed0047ad1a635cb0#egg=indexclient
git+https://git@github.com/uc-cdis/dictionaryutils.git@35899a878381eeb8ba91fd9124d285809d0063e8#egg=dictionaryutils
```
I'm not sure why either signpost or indexclient are there but definitely dictionaryutils should be added.

**Comment by philloooo** [comment]
I created a test branch in uc-cdis/gdcapi. Can you use that branch to test if things break 
it's failing right now https://travis-ci.com/uc-cdis/gdcapi/builds/63556592

**Comment by ProfOak** [comment]
Sure, the build is failing because requirements.txt isn't satisfying the new dependencies involved in sheepdog. sheepdog's setup.py says the install_requires cdiserrors, and a few other of our repositories like cdislogging. These packages are not in pypi and that's the only place pip knows to check. By default pip does not follow dependency_links in the dependency's setup.py. So they have to be added to gdcapi's requirements.txt as well. At least that will get the program to install.

I'll do this tomorrow when I'm in the office.

Also I thought about it some more and I think the most reasonable thing to do with dependency management is what you originally described. Everything in setup.py is not pinned to a version. These pins should come from requirements.txt. All applications importing these libraries should look to the library's requirements.txt to infer what versions to pin in it's requirements.txt. Every repository should have these two files:

* requirements.txt - Pinned versions of dependencies
* setup.py - No pins on dependencies- the minimum package information required to use the library

If there are tests in the repo then dev-requirements.txt should also exist. I also thought about renaming it to test-requirements.txt.

If you are installing a repo, it would look like this

```bash
$ pip install -r requirements.txt
$ python setup.py develop
$ pip install -r dev-requirements.txt```

Sorry for dumping everything here. I just read this message and didn't want to forget,

File: sheepdog/transactions/upload/entity.py [hunk 2]
```diff
@@ -163,6 +165,29 @@ def _get_node_create(self, skip_node_lookup=False):
                     type=EntityErrors.NOT_UNIQUE,
                 )
 
+        # Check to see if it's registered in dbGaP is is an exception
```
Comment (philloooo) [inline sheepdog/transactions/upload/entity.py]: verifying against dbgap should be optional
gen3 submission don't need to check with dbgap

File: setup.py [hunk 3]
```diff
@@ -1,26 +1,52 @@
...
+    ],
+    dependency_links=[
+        'git+https://git@github.com/uc-cdis/userdatamodel.git@cb7143c709a1173c84de4577d3e866318a2cc834#egg=userdatamodel',
+        'git+https://git@github.com/uc-cdis/cdis_oauth2client.git@0.1.2#egg=cdis_oauth2client',
+        'git+https://git@github.com/NCI-GDC/psqlgraph.git@1.2.0#egg=psqlgraph',
+        'git+https://git@github.com/uc-cdis/cdiserrors.git@0.0.4#egg=cdiserrors',
+        'git+https://git@github.com/uc-cdis/cdis-python-utils.git@0.2.2#egg=cdispyutils',
+        'git+https://git@github.com/uc-cdis/cdislogging.git@master#egg=cdislogging',
+        'git+https://git@github.com/uc-cdis/datamodelutils.git@0.2.0#egg=datamodelutils',
+        'git+https://git@github.com/uc-cdis/authutils.git@feat/api-373-separate-authutils#egg=authutils',
```
Comment (philloooo) [inline setup.py]: don't merge this one before authutils is merged and tagged with a version

**Pull request merged**